cdragontoolbox/hashes.py

```diff
@@ -205,38 +210,83 @@ def substitute_basenames(self):
         """Check all basenames in each subdirectory"""
 
         names = set(os.path.basename(p) for p in self.known.values())
-        dirs = self.directory_list()
+        dirs = list(self.directory_list())
+        unknown = self.unknown # global -> local for increased performance
         logger.info(f"substitute basenames: {len(names)} basenames, {len(dirs)} directories")
         for name in progress_iterator(sorted(names)):
-            self.check_iter(f"{d}/{name}" for d in dirs)
-
-    def _substitute_basename_words(self, words):
-        """Replace all words in known basename"""
+            for d in dirs:
+                h = xxh64(name).intdigest() # inline check() for increased performance
+                if h in unknown:
+                    self._add_known(h, name)
+
+    def _substitute_basename_words(self, paths, words, amount=1):
+        """Substitutes {amount} side by side words in all basenames
+        of all given paths (default=1 for simple 1 word substitution)."""
+
+        regex_part = ""
+        format_part = ""
+        for i in range(1, amount):
+            regex_part += r'([-_][^/_.-]+)'
```

Comment:
- `regex_part = r'...' * (amount-1)` (same for `format_part`)
- `.` is a valid separator used in some basenames

Reply:
If you decide to exclude the `.suffix.ext` trailing part used by some files, then you could simplify the regexp since you shouldn't need the `[^/]*\.[^/]+$` part.

Also, since you modified the regexp, you need to update the use of `m.group()` to: all parts are in the same group so you don't need to join them anymore.